trieste/models/normalization/model_wrappers.py

```diff
@@ -0,0 +1,232 @@
+        predict_function = getattr(super(), f'predict{predict_type}')
+        mean, covariance = predict_function(transformed_query_points, *args, **kwargs)
+        return (
+            self._observation_transformer.inverse_transform(mean),
+            self._observation_transformer.inverse_transform_variance(covariance)
+        )
+
+    def predict(
+        self, query_points: TensorType, *args: Any, **kwargs: Any
+    ) -> Tuple[TensorType, TensorType]:
+        """Wrap model's `predict` method to pass in a normalized dataset and return unnormalized outputs."""
+        return self._predict(query_points, *args, **kwargs)
+
+    def predict_joint(
+        self, query_points: TensorType, *args: Any, **kwargs: Any
+    ) -> Tuple[TensorType, TensorType]:
+        """Wrap model's `predict_joint` method to pass in a normalized dataset and return unnormalized outputs."""
+        return self._predict(query_points, *args, predict_type='_joint', **kwargs)
+
+    def sample(self, query_points: TensorType, *args: Any, **kwargs: Any) -> TensorType:
+        """Wrap the model's `sample` method to normalize the inputs before sampling, and unnormalize the outputs
+        before returning.
+
+        :param query_points: Unnormalized query points [..., D].
+        :param num_samples: Number of samples required, N.
+        :return: Unnormalized samples [N, ..., D]
+        """
+        transformed_query_points = self._query_point_transformer.transform(query_points)
+        samples = super().sample(transformed_query_points, *args, **kwargs)
+        return self._observation_transformer.inverse_transform(samples)
+
+    def predict_y(
+        self, query_points: TensorType, *args: Any, **kwargs: Any
+    ) -> Tuple[TensorType, TensorType]:
+        """Wrap model's `predict_y` method to pass in a normalized dataset and return unnormalized outputs."""
+        return self._predict(query_points, *args, predict_type='_y', **kwargs)
+
+    def update(self, dataset: Dataset, *args: Any, **kwargs: Any) -> None:
+        """Wrap the model's `update` method to pass in a normalized dataset. Optionally update normalization
+        and model parameters.
+        
+        :param dataset: Unnormalized dataset.
+        """
+        if self._update_parameters:
+            self._update_model_and_normalization_parameters(dataset)
+        transformed_dataset = self._transform_dataset(dataset)
+        super().update(transformed_dataset, *args, **kwargs)  # Will update data only.
+
+    def optimize(self, dataset: Dataset, *args: Any, **kwargs: Any) -> None:
+        """Wrap the model's `optimize` method to pass in a normalized dataset.
+        
+        :param dataset: Unnormalized dataset.
+        """
+        transformed_dataset = self._transform_dataset(dataset)
+        return super().optimize(transformed_dataset, *args, **kwargs)
+
+
+class GPRDataTransformWrapper(DataTransformWrapper, GaussianProcessRegression):
```

Comment:
huh, this is getting messier, would we need a wrapper per model? supporting this would be a nightmare...